cocotb/decorators.py

```diff
@@ -433,24 +433,43 @@ class test(_py_compat.with_metaclass(_decorator_helper, coroutine)):
     Used as ``@cocotb.test(...)``.
 
     Args:
-        timeout (int, optional):
-            value representing simulation timeout (not implemented).
+        timeout_time (int, optional):
+            value representing simulation timeout.
+        timeout_unit (str, optional):
+            unit of timeout value, see :class:`~cocotb.triggers.Timer` for more info.
         expect_fail (bool, optional):
             Don't mark the result as a failure if the test fails.
         expect_error (bool, optional):
             Don't mark the result as an error if an error is raised.
-            This is for cocotb internal regression use 
+            This is for cocotb internal regression use
             when a simulator error is expected.
         skip (bool, optional):
             Don't execute this test as part of the regression.
         stage (int, optional)
             Order tests logically into stages, where multiple tests can share a stage.
     """
-    def __init__(self, f, timeout=None, expect_fail=False, expect_error=False,
+    def __init__(self, f, timeout_time=None, timeout_unit=None,
+                 expect_fail=False, expect_error=False,
                  skip=False, stage=None):
+
+        if timeout_time is not None:
+            co = coroutine(f)
+
+            def timeout_wrapper(*args, **kwargs):
+                running_co = co(*args, **kwargs)
+                try:
+                    yield cocotb.triggers.with_timeout(running_co, self.timeout_time, self.timeout_unit)
+                except cocotb.result.SimTimeoutError:
+                    running_co.kill()
+                    raise
```

Comment:
Doesn't really matter, but I'd be inclined to propagate the return value here, by adding `ret = yield ...` and `raise ReturnValue(ret)` - just in case we ever use it, or someone abuses it for weird lifetime semantics